File: 0543-diameter-of-binary-tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let  diameter  = [0];
    height(root, diameter);
    return  diameter[0];
};

function height(root, diameter){

    if(root == null) return 0;

    let leftHeight = height(root.left, diameter);
    let rightHeight = height(root.right, diameter);

    diameter[0] =  Math.max( diameter[0], leftHeight+rightHeight );

    return Math.max(leftHeight , rightHeight ) + 1;

}

File: 0543-diameter-of-binary-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
int height(TreeNode* root, int diameter[]){

    if(root == NULL) return 0;

    int leftHeight = height(root->left, diameter);
    int rightHeight = height(root->right, diameter);

    diameter[0] =  max( diameter[0], leftHeight+rightHeight );

    return max(leftHeight , rightHeight ) + 1;

}
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter[1] = {0};
        height(root, diameter);
        return diameter[0];
    }
};
